feat: Add Azure OpenAI chat completions provider

nfuller286 opened the pull request:
This commit introduces support for integrating Azure OpenAI as a new chat completion source within SillyTavern. The implementation specifically navigates Azure's reliance on deployment names, differing from standard model selection.

Includes:

- **New Feat / UI:** Dedicated settings in `public/index.html` for Azure parameters (Base URL, Deployment Name, API Version, API Key) and a new `public/img/azure_openai.svg` icon.
- **Frontend Logic (`public/scripts/openai.js`):** Manages Azure OpenAI settings and dynamic URL generation. The 'Connect' button triggers a `GET /models` endpoint call solely for API configuration validation (key, URL, version). Upon successful validation, a subsequent `POST` request is made to the AI, and the active model is extracted from its response payload, then populated for UI display (no direct user model selection).
- **Backend Logic (`src/endpoints/backends/chat-completions.js`):** Implements `sendAzureOpenAIRequest` to proxy requests, handling Azure’s unique authentication and URL structures. Enhanced the `/status` endpoint for Azure-specific validation.
- **Secret Management:** Secure handling of Azure OpenAI API keys in `public/scripts/secrets.js` and `src/endpoints/secrets.js`.
- **Autoconnect:** `public/scripts/RossAscends-mods.js` modified to enable automatic connection for Azure OpenAI if an API key is present.

Impact:
- Users can now connect to and utilize Azure OpenAI services.
- Requires new configuration details in the UI.
- Enhances API routing and validation with Azure-specific behavior.
- Navigates model handling: model name is *derived dynamically* from AI response, not directly selected, aligning with internal SillyTavern structure.

<!-- Put X in the box below to confirm -->

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

nfuller286 committed:
feat: Add Azure OpenAI chat completions provider

This commit introduces comprehensive support for integrating Azure OpenAI as a new chat completion source within SillyTavern. The implementation specifically navigates Azure's reliance on deployment names, differing from standard model selection.

Includes:

- **New Feat / UI:** Dedicated settings in `public/index.html` for Azure parameters (Base URL, Deployment Name, API Version, API Key) and a new `public/img/azure_openai.svg` icon.
- **Frontend Logic (`public/scripts/openai.js`):** Manages Azure OpenAI settings and dynamic URL generation. The 'Connect' button triggers a `GET /models` endpoint call solely for API configuration validation (key, URL, version). Upon successful validation, a subsequent `POST` request is made to the AI, and the active model is extracted from its response payload, then populated for UI display (no direct user model selection).
- **Backend Logic (`src/endpoints/backends/chat-completions.js`):** Implements `sendAzureOpenAIRequest` to proxy requests, handling Azure’s unique authentication and URL structures. Enhanced the `/status` endpoint for Azure-specific validation.
- **Secret Management:** Secure handling of Azure OpenAI API keys in `public/scripts/secrets.js` and `src/endpoints/secrets.js`.
- **Autoconnect:** `public/scripts/RossAscends-mods.js` modified to enable automatic connection for Azure OpenAI if an API key is present.

Impact:
- Users can now connect to and utilize Azure OpenAI services.
- Requires new configuration details in the UI.
- Enhances API routing and validation with Azure-specific behavior.
- Navigates model handling: model name is *derived dynamically* from AI response, not directly selected, aligning with internal SillyTavern structure.

nfuller286 commented:
The ESlint is pulling an error on a portion of the file that I didn't modify. It seems upset that I modified the file at all. 

Not sure if I am approaching adding the Azure OpenAI chat completion source the best way or not. The 3 main differences with the Azure OpenAI and regular OpenAI api is that:
1. Azure OpenAI wants an api-key authorization instead of bearer (though it will accept a bearer auth with the api key).
2. Azure OpenAI doesn't accept a 'model' parameter in the POST request. Instead it uses a deployment name that is determined by the URL. 
3. Azure OpenAI requires an api version parameter to be tacked onto the request url. 

I added a section to the /status backend that extracts the model used in the deployment and returns it to the frontend to be used for the various parts of ST that need a model parameter.

Cohee1207 commented on public/index.html:
```diff
@@ -2802,10 +2802,10 @@ <h4 class="margin0" data-i18n="Chat Completion Source">
                             <optgroup>
                                 <option value="ai21">AI21</option>
                                 <option value="aimlapi">AI/ML API</option>
+                                <option value="azure_openai">Azure OpenAI</option>
```
I think we can just call it `azure`

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
                             </select>
                         </div>
-                        <div id="prompt_post_processing_form">
+                        <form id="azure_openai_settings" data-source="azure_openai">
```
This is not actually a form. Change the block tag to `div`.

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                                <input id="api_key_azure_openai" data-setting="api_key_azure_openai" class="text_pole flex1" type="password" autocomplete="off" placeholder="Enter your Azure API Key">
+                                <div title="Manage API keys" data-i18n="[title]Manage API keys" class="menu_button fa-solid fa-key fa-fw manage-api-keys" data-key="api_key_azure_openai"></div>
+                            </div>
+                            <div class="neutral_warning" data-i18n="For privacy reasons, your API key will be hidden after you click 'Connect'.">
```
This block lacks a data attribute, so it will be permanently displayed instead of just when the key input is populated

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                            <!-- Azure API Key -->
+                            <h4><span data-i18n="Azure API Key">Azure API Key</span></h4>
+                            <div class="flex-container">
+                                <input id="api_key_azure_openai" data-setting="api_key_azure_openai" class="text_pole flex1" type="password" autocomplete="off" placeholder="Enter your Azure API Key">
```
Placeholder will be replaced by secrets.js script. Remove it from here

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                                    <option value="2024-10-21">2024-10-21</option>
+                                </select>
+                            </div>
+                             <div>
```
I don't think this text brings any significant value.

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                            <div class="flex-container">
+                                <input id="azure_deployment_name" data-setting="azure_deployment_name" class="text_pole wide100p" type="text" placeholder="your-deployment-name">
+                            </div>
+                            <div>
```
Move this to a `title` attribute, or remove it.

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                            <div>
+                                <small data-i18n="The underlying model of your deployment. This is detected automatically when you connect.">The underlying model of your deployment. This is detected automatically when you connect.</small>
+                            </div>
+                        </form>                        <div id="prompt_post_processing_form">
```
Fix the formatting.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -329,6 +332,12 @@ export const settingsToUpdate = {
+    azure_deployment_name: ['#azure_deployment_name', 'azure_deployment_name', false, true],
+    azure_api_version: ['#azure_api_version', 'azure_api_version', false, true],
+    azure_openai_model: ['#azure_openai_model', 'azure_openai_model', false, true],
+
```
Remove empty lines

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,65 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                            <!-- Model Name (Read-only) -->
+                            <h4><span data-i18n="Model Name">Model Name</span></h4>
+                            <div class="flex-container">
+                                <input id="azure_openai_model" data-setting="azure_openai_model" class="text_pole wide100p" type="text" readonly placeholder="Click 'Connect' to fetch model name">
```
`azure_openai_model` is an input, but you're adding `option` elements to it. Why not make it a `select`?

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,108 @@ async function sendAimlapiRequest(request, response) {
+
+    // --- Construct Azure API URL ---
+    // Ensure no double slashes if base_url already ends with one
+    const sanitizedBaseUrl = AzureOpenAIData.azure_base_url.endsWith('/') ? AzureOpenAIData.azure_base_url.slice(0, -1) : AzureOpenAIData.azure_base_url;
```
I think we already have a utility method that removes a trailing slash somewhere in the src/util.js module

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1675,7 +1867,6 @@ router.post('/generate', function (request, response) {
         bodyParams = {
             reasoning_effort: request.body.reasoning_effort,
         };
-        throw new Error('This provider is temporarily disabled.');
```
Revert this back.

Cohee1207 replied:
The comment is still relevant.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1254,7 +1356,6 @@ router.post('/status', async function (request, statusResponse) {
         apiUrl = API_COMETAPI;
         apiKey = readSecret(request.user.directories, SECRET_KEYS.COMETAPI);
         headers = {};
-        throw new Error('This provider is temporarily disabled.');
```
Revert this back.

Cohee1207 replied:
The comment is still relevant.

Cohee1207 commented on public/index.html:
```diff
@@ -2802,10 +2802,10 @@ <h4 class="margin0" data-i18n="Chat Completion Source">
+                                <option value="azure_openai">Azure OpenAI</option>
                                 <option value="claude">Claude</option>
                                 <option value="cohere">Cohere</option>
-                                <!-- Temporarily disabled. -->
```
Revert this back.

Cohee1207 replied:
The comment is still relevant.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -3723,12 +3799,20 @@ async function getStatusOpen() {
         chat_completion_source: oai_settings.chat_completion_source,
     };
 
+
```
Remove blank lines

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -3748,6 +3832,16 @@ async function getStatusOpen() {
+        data.azure_base_url = oai_settings.azure_base_url;
+        data.azure_deployment_name = oai_settings.azure_deployment_name;
+        data.azure_api_version = oai_settings.azure_api_version;
+        data.testPayload = {
```
Where is this used? I don't see any more references to `testPayload` anywhere.

Cohee1207 commented on public/scripts/secrets.js:
```diff
@@ -194,6 +197,14 @@ export function resolveSecretKey() {
             }
         }
+        // ADD THIS BLOCK FOR AZURE OPENAI
+        if (chatCompletionSource === chat_completion_sources.AZURE_OPENAI) {
```
This is not needed. Both secret_keys and chat_completion_sources have the same AZURE_OPENAI key.
Also the block is duplicated twice.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1299,6 +1400,96 @@ router.post('/status', async function (request, statusResponse) {
+                    console.warn('Failed to read error text from Azure response:', e);
+                }
+                console.warn('Azure OpenAI GET /models failed:', apiConfigTest.status, apiConfigTest.statusText, errText || '');
+                const message =
```
Place error messages in a map object for simpler code.

Cohee1207 replied:
The comment is still relevant. Also just reading the status text from the response will be enough.

Cohee1207 commented:
What about the following features?

1. Tool calling
2. Prompt image inlining
3. Structured (JSON) output

Also do you plan to support the integration in the following extensions?

1. Image Captioning
2. Image Generation
3. TTS
4. Vector Storage

nfuller286 committed:
fixed the html issues, openai.ujs and secrets.js

	modified:   public/index.html#
    modified:   public/scripts/openai.js
	modified:   public/scripts/secrets.js

nfuller286 reviewed (commented):
> What about the following features?
> 
> 1. Tool calling
> 2. Prompt image inlining
> 3. Structured (JSON) output
> 
> Also do you plan to support the integration in the following extensions?
> 
> 1. Image Captioning
> 2. Image Generation
> 3. TTS
> 4. Vector Storage 

Thanks for checking this out. I'm making some changes, but am having trouble with debugging in VScode.

nfuller286 reviewed (commented):
> > What about the following features?
> > 
> > 1. Tool calling
> > 2. Prompt image inlining
> > 3. Structured (JSON) output
> > 
> > Also do you plan to support the integration in the following extensions?
> > 
> > 1. Image Captioning
> > 2. Image Generation
> > 3. TTS
> > 4. Vector Storage 
> 
> Thanks for checking this out. I'm making some changes, but am having trouble with debugging in VScode.

Scratch that. I figured it out. Was being dense.

nfuller286 committed:
Extends Azure OpenAI with advanced capabilities

Enables advanced features such as function calling, image inlining, and model reasoning in the UI for Azure OpenAI.

Refactors the backend to support structured output (JSON mode), native thinking, and reasoning effort for compatible Azure OpenAI models. Adds support for logprobs, multiple responses (`n`), and seed, aligning with standard OpenAI behavior.

Improves request handling with a retry mechanism for rate limits and more robust error reporting. Ensures correct parameter handling for model-specific features and conflicts, such as when native thinking is enabled. Optimizes the Azure OpenAI status probe for efficiency.
	modified:   public/index.html
	modified:   public/scripts/openai.js
	modified:   src/endpoints/backends/chat-completions.js

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+
+    // Define the list of models that support these advanced features
+    const reasoningEffortModels = [
+        'o1', 'o3-mini', 'o3-mini-2025-01-31', 'o4-mini', 'o4-mini-2025-04-16',
```
Can we somehow avoid the list, please?

nfuller286 replied:
Yep. Simplified this whole area.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+        }
+    }
+
+    // *** START: ADDED LOGIC FOR JSON MODE ***
```
Remove the start/end comments, please. It's already evident where the block starts and ends.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+    // *** START: ADDED LOGIC FOR JSON MODE ***
+    // Handle Structured Output (JSON Mode) by translating the custom `json_schema` object.
+    if (request.body.json_schema) {
+        console.info('Enabling JSON Mode via structured output schema.');
```
Remove the console log.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+        if (includeReasoning) {
+            // gpt-5 models do not support stream_options.include_thoughts
+            if (!modelName.startsWith('gpt-5')) {
+                console.info(`Enabling native 'thoughts' API feature for model: ${modelName}`);
```
Remove the console log.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+        if (request.body.reasoning_effort) {
+            console.info(`Setting 'reasoning_effort' for model: ${modelName}`);
+            const reasoningEffortMap = {
+                min: 'minimal',
```
Models before gpt-5 don't support minimal reasoning effort.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+        }
+    }
+
+    async function executeRequest(url, fetchConfig, retries = 3, timeout = 5000) {
```
Please don't retry on 429, I already removed that logic from core code. Remove these inner functions and execute fetch directly.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1299,6 +1464,96 @@ router.post('/status', async function (request, statusResponse) {
             console.error('Error fetching Google AI Studio models:', error);
             return statusResponse.send({ error: true, bypass: true, data: { data: [] } });
         }
+
```
Remove the added empty line

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1299,6 +1464,96 @@ router.post('/status', async function (request, statusResponse) {
+                body: JSON.stringify(modelPayload),
+            });
+
+            // Parse response safely even if server mislabels content-type
```
Why would it? And it shouldn't even forbid running json() on the response even if content-type header is mislabeled.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1299,6 +1464,96 @@ router.post('/status', async function (request, statusResponse) {
+        }
+
+        // 2) Build URLs
+        const base = azure_base_url.endsWith('/') ? azure_base_url.slice(0, -1) : azure_base_url;
```
We have a shared utility function (trimTrailingSlash) for this specific purpose.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1195,6 +1195,172 @@ async function sendAimlapiRequest(request, response) {
+    const validKeys = ['messages', 'temperature', 'frequency_penalty', 'presence_penalty', 'top_p', 'max_tokens', 'stream', 'logit_bias', 'stop', 'n', 'logprobs', 'seed', 'tools', 'tool_choice'];
+    const apiRequestBody = /** @type {any} */ ({});
+    for (const key of validKeys) {
+        if (request.body[key] !== undefined) {
```
I think it's better to use `Object.hasOwn` to check if the presence of a key.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -6162,6 +6299,26 @@ export function initOpenAI() {
+
+    $('#azure_deployment_name').on('input', function () {
+        oai_settings.azure_deployment_name = String($(this).val());
+        updateAzureCompleteUrl(); // Keep this to update the URL preview
```
Remove the comment

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -6162,6 +6299,26 @@ export function initOpenAI() {
+    // Add input listeners for Azure OpenAI settings to update oai_settings and save
+    $('#azure_base_url').on('input', function () {
+        oai_settings.azure_base_url = String($(this).val());
+        updateAzureCompleteUrl(); // Keep this to update the URL preview
```
Remove the comment

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -6162,6 +6299,26 @@ export function initOpenAI() {
+
+    $('#azure_api_version').on('input change', function () {
+        oai_settings.azure_api_version = String($(this).val());
+        updateAzureCompleteUrl(); // Keep this to update the URL preview
```
Remove the comment

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -6162,6 +6299,26 @@ export function initOpenAI() {
+        updateAzureCompleteUrl(); // Keep this to update the URL preview
+        saveSettingsDebounced();
+    });
+
```
Remove the added empty line

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -5808,6 +5925,26 @@ function updateVertexAIServiceAccountStatus(isValid = false, message = '') {
+
+    $('#azure_complete_url').val(completeUrl);
+}
+
```
Remove the added empty line

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -5808,6 +5925,26 @@ function updateVertexAIServiceAccountStatus(isValid = false, message = '') {
+/**
+ * Updates the #azure_complete_url input field based on Azure OpenAI settings.
+ */
+function updateAzureCompleteUrl() {
```
Is there any value in showing this URL in the UI? I don't think there could be any confusion in where exactly the response will be sent. This is safe to remove.

nfuller286 replied:
Beyond me just enjoying watch it update in real time, I suppose not. Got rid of it and got rid of the html.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -5370,6 +5479,9 @@ function toggleChatCompletionForms() {
         $('#model_cometapi_select').trigger('change');
     }
 
```
Place the empty line after the added block, not before.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -5299,6 +5398,16 @@ async function onConnectButtonClick(e) {
+        // If a new key has been entered in the text box, save it before connecting.
+        if (api_key_azure_openai.length) {
+            await writeSecret(SECRET_KEYS.AZURE_OPENAI, api_key_azure_openai);
+        }
```
The logic that stops the function if no key is provided is missing
```js

        if (!api_key_azure_openai && !secret_state[SECRET_KEYS.AZURE_OPENAI]) {
            console.log('No secret key saved for Azure OpenAI');
            return;
        }
```

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4812,17 +4910,18 @@ async function onModelChange() {
         oai_settings.temp_openai = Math.min(oai_max_temp, oai_settings.temp_openai);
         $('#temp_openai').attr('max', oai_max_temp).val(oai_settings.temp_openai).trigger('input');
     }
-
```
Re-add the line back.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4812,17 +4910,18 @@ async function onModelChange() {
         $('#openai_max_context').val(oai_settings.openai_max_context).trigger('input');
 
-        $('#openai_reverse_proxy').attr('placeholder', 'https://api.openai.com/v1');
+        if (oai_settings.chat_completion_source == chat_completion_sources.OPENAI) {
```
This block is not visible when the source is not OpenAI, so this added condition is not needed.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4812,17 +4910,18 @@ async function onModelChange() {
     }
 
-    if (oai_settings.chat_completion_source == chat_completion_sources.OPENAI) {
+    if (oai_settings.chat_completion_source == chat_completion_sources.AZURE_OPENAI || oai_settings.chat_completion_source == chat_completion_sources.OPENAI) {
```
Please use an array.includes check for cleaner code
```js
    if ([chat_completion_sources.OPENAI, chat_completion_sources.AZURE_OPENAI].includes(oai_settings.chat_completion_source)) {
```

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4740,6 +4833,11 @@ async function onModelChange() {
     }
 
+    if ($(this).is('#azure_openai_model')) {
+        console.log('Azure OpenAI model changed to', value);
```
Is it possible that the model id will be empty? If yes, then add a return like in a block above, otherwise it will be impossible to auto-select the context size, etc.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2151,6 +2207,14 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
     };
 
+    // Add Azure-specific settings to generate_data if Azure OpenAI is selected
+    if (isAzureOpenAI) { // Use the new isAzureOpenAI constant
```
Remove the added comment

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2151,6 +2207,14 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
         'custom_prompt_post_processing': oai_settings.custom_prompt_post_processing,
     };
 
+    // Add Azure-specific settings to generate_data if Azure OpenAI is selected
```
Remove the added comment

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2151,6 +2207,14 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
+        generate_data.azure_base_url = oai_settings.azure_base_url;
+        generate_data.azure_deployment_name = oai_settings.azure_deployment_name;
+        generate_data.azure_api_version = oai_settings.azure_api_version;
+        // The 'model' field will be present but is handled (deleted) by the backend for Azure.
```
Remove the added comment. Also it's not, it is passed in the code above.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1928,6 +1945,43 @@ function saveModelList(data) {
+            $el.val(oai_settings.azure_openai_model || modelId).trigger('input').trigger('change');
+        }
+    }
+
```
Remove added empty line

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1928,6 +1945,43 @@ function saveModelList(data) {
+        }
+
+        // Populate the target element. Support both <input> and <select>.
+        if ($el.is('select')) {
```
It's always a select (as evident by the HTML), so "else" block will never execute.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1928,6 +1945,43 @@ function saveModelList(data) {
 
         $('#model_cometapi_select').val(oai_settings.cometapi_model).trigger('change');
     }
+    // --- Azure OpenAI: populate #azure_openai_model from the single returned item ---
```
Replace this comment with an empty line.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -1928,6 +1945,43 @@ function saveModelList(data) {
+    // --- Azure OpenAI: populate #azure_openai_model from the single returned item ---
+    if (oai_settings.chat_completion_source == chat_completion_sources.AZURE_OPENAI) {
+        const $el = $('#azure_openai_model');
+        if ($el.length === 0) {
```
Jquery gracefully handles missing DOM elements. So you're free to remove this conditional return.

Cohee1207 commented on public/index.html:
```diff
@@ -3720,7 +3720,61 @@ <h4 data-i18n="Moonshot AI Model">Moonshot AI Model</h4>
+                            <div>
+                                <small data-i18n="The underlying model of your deployment. This is detected automatically when you connect.">The underlying model of your deployment. This is detected automatically when you connect.</small>
+                            </div>
+                        </div>                        <div id="prompt_post_processing_form">
```
Fix this line: add a line break after a closing tag and before the opening tag for prompt_post_processing_form

Cohee1207 commented on public/index.html:
```diff
@@ -2802,10 +2802,10 @@ <h4 class="margin0" data-i18n="Chat Completion Source">
                             <optgroup>
                                 <option value="ai21">AI21</option>
                                 <option value="aimlapi">AI/ML API</option>
+                                <option value="azure_openai">Azure</option>
```
I actually meant changing the value to `azure`, not the text label, but actually on the second thought that's fine. So you can rename it back to "Azure OpenAI"
```suggestion
                                <option value="azure_openai">Azure OpenAI</option>
```

Cohee1207 reviewed (changes requested):
Please sync your branch with the latest staging before fixing any of the comments.

nfuller286 committed:
Merge branch 'develop' into feature/azure-openai

nfuller286 committed:
PR fixes
Simplification
Removed reasoning logic from backend
	modified:   src/endpoints/backends/chat-completions.js

	modified:   src/endpoints/backends/chat-completions.js

nfuller286 committed:
Fix PR comments
Removed the front end compiled url UI element and related.
Misc simplifications
	modified:   public/index.html
	modified:   public/scripts/openai.js
	modified:   src/endpoints/backends/chat-completions.js

nfuller286 committed:
Merge branch 'develop' into feature/azure-openai

nfuller286 committed:
Merge branch 'fix/pr-final-review' into feature/azure-openai

nfuller286 committed:
Fixed accidental api temporary disabled
	modified:   public/index.html

nfuller286 commented:
Alright, I believe all the comments have been addressed. I also simplified the backend.

Cohee1207 committed:
Don't transfer status code verbatim

Cohee1207 committed:
Fix formatting

Cohee1207 committed:
Enable tool calling

Cohee1207 committed:
Fix logo coloration

Cohee1207 committed:
Move model arrays to shared constants

Cohee1207 committed:
Fix capitalization

Cohee1207 committed:
Remove obsolete comment

Cohee1207 committed:
Improve response schema parameter format

Cohee1207 committed:
Fix tokenizer model selection

Cohee1207 reviewed (approved):
Tested, works.
